src/mlpack/methods/hdbscan/hdbscan_impl.hpp

```diff
@@ -0,0 +1,763 @@
+template<typename MatType>
+void HDBSCAN<NeighborSearch,
+    MetricType,
+    TreeType>::
+Cluster(const MatType& data,
+        arma::Row<size_t>& assignments)
+{
+  // genearting a matrix with the values of dcore
+  arma::Mat<size_t> neighbors;
+  arma::mat distances;
+  neighborSearch.Train(data);
+  neighborSearch.Search(minPoints, neighbors, distances);
+  arma::mat dataWithDcore = arma::conv_to<arma::mat>::from(data);
+  dataWithDcore.resize(dataWithDcore.n_rows+1, dataWithDcore.n_cols);
+  dataWithDcore.row(dataWithDcore.n_rows-1) = distances.row(distances.n_rows-1);
+
+  // computing the mst of the given data and sort the obtained result
+  emst::DualTreeBoruvka<MetricType, arma::mat, TreeType>
+                       dtb(dataWithDcore, false, hMetric);
+  arma::mat resultsMST;
+  dtb.ComputeMST(resultsMST);
+
+  // create single linkage tree from the MST obtained.
+  arma::mat singleLinkageTree;
+  SingleLinkageTreeClustering(resultsMST, singleLinkageTree);
+
+  // Condense the obtained tree
+  arma::mat condensedTree;
+  CondenseTree(singleLinkageTree, condensedTree, minPoints);
+
+  // get clusters
+  arma::Mat<size_t> result;
+  GetClusters(condensedTree, result);
+
+  // save results obtained
+  assignments = result;
+}
+
+//  A class to help merge two clusters to form a
+// single linkage tree.
+class UnionForSingleLinkageClustering
+{
+  arma::Col<size_t> parentVector;
+  arma::Col<size_t> sizeVector;
+  size_t nextLabel;
+
+ public:
+  UnionForSingleLinkageClustering(size_t n)
+  {
+    // Initialize the elments
+    // There would be a max of 2*n-1 clusters
+    // formed after completion
+    // Initally all n points form a cluster
+    // After first merging there are
+    // (n-2) old clusters and 1 new cluster
+    // This goes on until at last there is one cluster
+    // (n-1) + n = (2n-1)
+    parentVector.set_size(2*n-1);
+    parentVector.fill(SIZE_MAX);
+    sizeVector.set_size(2*n-1);
+    for (size_t i = 0; i < n; i++)
+      sizeVector(i) = 1;
+    for (size_t i = n; i < 2*n-1; i++)
+      sizeVector(i) = 0;
+    nextLabel = n;
+  }
+  void UnionOfTwoClusters(size_t m, size_t n)
+  {
+    // Each union will form a new cluster
+    // with size as sum of its children
+    sizeVector(nextLabel) = sizeVector(m) + sizeVector(n);
+    parentVector(m) = nextLabel;
+    parentVector(n) = nextLabel;
+    nextLabel++;
+  }
+  size_t Find(size_t n)
+  {
+    // finds the parent
+    size_t root = n;
+    while (parentVector(root) != SIZE_MAX)
+      root = parentVector(root);
+    while (parentVector(n) != root)
+    {
+      size_t temp = parentVector(n);
+      parentVector(n) = root;
+      n = temp;
+      if (n == SIZE_MAX)
+        break;
+    }
+      return root;
+  }
+  size_t GetSize(size_t i)
+  {
+    return sizeVector(i);
+  }
+};
+
+/**
+ * Converts the input (sorted MST) to a single linkage tree.
+ * Each row of the output contaisn 3 values
+ * (cluster1 cluster2 d size)
```

Comment:
Don't be afraid to use multiple types here.  I would suggest using one object of type arma::Mat<size_t> to hold cluster1 and cluster2 and size, and then a third MatType to hold the distances between the clusters.